### ISSUE TITLE
InMemory: Reading value from server may need to apply default

### ISSUE THREAD

**Issue by smitpatel** [dotnet/efcore#16575]
```
IQueryable<int> Select(
|__ source: DbSet<Customer>, 
|__ selector: (Customer Param_0) => (int)Nullable<int> FirstOrDefault(IQueryable<Nullable<int>> Select(
    |__ source: IQueryable<string> Distinct(IQueryable<string> Select(
    |   |__ source: IOrderedQueryable<Order> OrderBy(
    |   |   |__ source: IQueryable<Order> Where(
    |   |   |   |__ source: DbSet<Order>, 
    |   |   |   |__ predicate: (Order o) =>  ?= Property(Param_0, "CustomerID") == Property(o, "CustomerID") =? ), 
    |   |   |__ keySelector: (Order Param_0) => Param_0.OrderID), 
    |   |__ selector: (Order Param_0) => Param_0.CustomerID)), 
    |__ selector: (string Param_0) => (Nullable<int>)Param_0.Length)))
```
Above pass for relational since while reading the value we try to read int value from null column giving us back default which is 0. It fails for in memory because we read null and try to cast to int.

**Comment by roji** [comment]
Simplified repro:

```c#
using (var context = CreateContext())
{
    var x = context.Customers
        .Select(c => context.Orders
            .Where(o => o == local)
            .Select(o => o.CustomerID)
            .FirstOrDefault().Length)
        .ToList();
}
```

**Comment by ajcvickers** [comment]
Closed in favor of #16963
